### ISSUE TITLE
Unexpected behaviour in 2D Delaunay triangulation

### ISSUE THREAD

**Issue by keeeal** [LucaCiucci/qhull-rs#1]
Thanks for writing this crate! It is exactly what I have needed for a really long time 🙏 

I am using it to perform a Delaunay triangulation in 2D. The following example has only 4 points:

```rs
fn main() {
    let mut rng = rand::thread_rng();
    let mut points = Vec::new();

    for _ in 0..4 {
        points.push([
            rng.gen_range(-1.0..1.0),
            rng.gen_range(-1.0..1.0),
        ]);
    }

    let qh = Qh::new_delaunay(points).unwrap();

    for simplex in qh.simplices() {
        println!("---");
        for vertex in simplex.vertices().unwrap().iter() {
            println!("{:?}", vertex.dim());
        }
    }
}
```

The output of this program is

```
---
3
3
3
---
3
3
3
---
3
3
3
---
3
3
3
```

which tells me the output consists of four triangles in R3, i.e. a tetrahedron.

I expected the output to have either 2 or 3 triangles, and for it to be in R2 since the input points were 2D.

Is this expected behaviour? Am I using it correctly?

**Comment by keeeal** [comment]
Further, looking at the example in `qhull-rs/examples`, given the following input points:

https://github.com/LucaCiucci/qhull-rs/blob/bf9ebe5633af5e27985d7c7be2b33b5cc82be420/examples/delaunay.rs#L5

I would expect the output **not** to include the triangle specified by indices `[0, 1, 2]` since it is the hull of the shape.

https://github.com/LucaCiucci/qhull-rs/blob/bf9ebe5633af5e27985d7c7be2b33b5cc82be420/examples/delaunay.rs#L35

Compare this with the equivalent using [scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.Delaunay.html), which also uses qhull under the hood:

```py
>>> from scipy.spatial import Delaunay
>>> qh = Delaunay([[0.0, 0.0], [1.0, 0.0], [0.0, 1.0], [0.25, 0.25]])
>>> qh.simplices
array([[1, 3, 0],
       [3, 2, 0],
       [2, 3, 1]], dtype=int32)
```

**Comment by LucaCiucci** [comment]
Hi @keeeal, thank you for reporting this issue.  
`qhull-rs`  is definitely wrong... I'll take a look into it ASAP.
Can you run a `cargo test`  on your machine?

**Comment by LucaCiucci** [comment]
Qhull does not provide a way to create a Delaunay triangulation directly, this is actually done by computing the convex hull of a paraboloid in a n+1 dimension. Maybe I forgot to remove the extra triangles.  
This doesn't explain the problem with your first example though...

**Comment by LucaCiucci** [comment]
> Can you run a `cargo check`  on your machine?

`cargo test` sorry

**Comment by keeeal** [comment]
> Can you run a cargo test on your machine?

```
cargo test 
    Blocking waiting for file lock on build directory
   Compiling qhull-sys v0.3.1 (C:\Users\j\qhull-rs\qhull-sys)
   Compiling qhull v0.3.1 (C:\Users\j\qhull-rs)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 30.74s
     Running unittests src\lib.rs (target\debug\deps\qhull-cc8244c82dcc33fb.exe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src\bin\qconvex.rs (target\debug\deps\qconvex-0df7d210141e105a.exe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src\bin\qdelaunay.rs (target\debug\deps\qdelaunay-244231b208b2e13c.exe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src\bin\qhalf.rs (target\debug\deps\qhalf-9e1bb28038a94e7a.exe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src\bin\qhull.rs (target\debug\deps\qhull-be4e14cf392e05d0.exe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src\bin\qvoronoi.rs (target\debug\deps\qvoronoi-1625aa7e44d8a9db.exe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src\bin\rbox.rs (target\debug\deps\rbox-541c7f5bbc2151a6.exe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests qhull

running 14 tests
test src\builder.rs - builder::QhBuilder::build (line 142) ... ok
test src\builder.rs - builder::QhBuilder (line 14) ... ok
test src\builder.rs - builder::QhBuilder::compute (line 94) ... ok
test src\builder.rs - builder::QhBuilder::build_managed (line 231) ... ok
test src\builder.rs - builder::QhBuilder::with_configure (line 299) ... ok
test src\error.rs - error::QhError<'a>::into_static (line 92) ... ok
test src\builder.rs - builder::QhBuilder::build_from_iter (line 262) ... ok
test src\examples.rs - examples (line 35) ... ok
test src\helpers.rs - helpers::collect_coords (line 37) ... ok
test src\examples.rs - examples (line 6) ... ok
test src\lib.rs - (line 55) ... ok
test src\helpers.rs - helpers::prepare_delaunay_points (line 83) ... ok
test src\lib.rs - Qh<'a>::try_on_qh (line 121) ... ok
test src\lib.rs - Qh<'a>::try_on_qh (line 138) ... ok

test result: ok. 14 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2.44s
```

**Comment by keeeal** [comment]
All passed ✅

**Comment by keeeal** [comment]
Another issue (or maybe a symptom of the same issue) is that the Delaunay triangulation seems to always contain all possible triangles. Consider the following points:

```
[[0.0, 0.0], [0.0, 1.1], [1.0, 0.0], [1.0, 1.0]]
```

#### Expected triangulation:

![scipy-spatial-Delaunay-1_00_00](https://github.com/user-attachments/assets/9cf1be68-ed14-43fc-81c5-b2d2d61ab85a)

I would only expect there to be two simplices in the output, but the following assertion still passes:

```rs
use qhull::Qh;
use std::error::Error;

fn main() -> Result<(), Box<dyn Error>> {
    let qh = Qh::new_delaunay([[0.0, 0.0], [0.0, 1.1], [1.0, 0.0], [1.0, 1.0]])?;

    let mut simplices = qh
        .simplices()
        .filter(|f| !f.is_sentinel())
        .map(|f| {
            f.vertices()
                .unwrap()
                .iter()
                .map(|v| v.id() - 1)
                .collect::<Vec<_>>()
        })
        .collect::<Vec<_>>();

    simplices.iter_mut().for_each(|s| s.sort());
    simplices.sort();
    assert_eq!(
        simplices,
        vec![vec![0, 1, 2], vec![0, 1, 3], vec![0, 2, 3], vec![1, 2, 3],]  // This should fail!
    );

    Ok(())
}
```

Again comparing this to scipy's output:

```py
>>> from scipy.spatial import Delaunay
>>> qh = Delaunay([[0.0, 0.0], [0.0, 1.1], [1.0, 0.0], [1.0, 1.0]])
>>> qh.simplices
array([[2, 3, 0],
       [3, 1, 0]], dtype=int32)
```

... two triangles, as expected.

**Comment by keeeal** [comment]
From the qhull docs:

```
The underlying structure is the lower facets of a convex hull in dimension d+1
```

Do the upper facets need filtering out - presumably by checking the direction of their normal?

**Comment by LucaCiucci** [comment]
Thank you for the details!  
I think this has something to do with setting the right options but there is something really wrong with this crate...

I'll leave here some notes.

The augmented set produced by [this crate](https://github.com/LucaCiucci/qhull-rs/blob/3ef56b498bac225d37bad962d4ab86c03afea954/src/helpers.rs#L134) is different from the one produced by [qhull](https://github.com/qhull/qhull/blob/89341b937a90ae8dc19e13a2bd4225d3c9abe4c2/src/libqhull_r/io_r.c#L3916)  but this does not seem to influence the resulting triangulation and should just be an enhancement.  
Example:
```
- [0.000000, 0.000000, 0.781250 ]
- [1.000000, 0.000000, 2.281250 ]
- [0.000000, 1.000000, 2.281250 ]
- [0.250000, 0.250000, 0.031250 ]
```

In `Qh::new_delaunay` there is a [TODO](https://github.com/LucaCiucci/qhull-rs/blob/3ef56b498bac225d37bad962d4ab86c03afea954/src/lib.rs#L73), the reference implementation to look at is [here](https://github.com/qhull/qhull/blob/89341b937a90ae8dc19e13a2bd4225d3c9abe4c2/src/qdelaunay/qdelaun_r.c).  
Here is the configuration that `Qh::new_delaunay` is currently producing:
<details>

<summary>qhT debug</summary>

```ron
qhT {
    ALLpoints: 0,
    ALLOWshort: 0,
    ALLOWwarning: 0,
    ALLOWwide: 0,
    ANGLEmerge: 0,
    APPROXhull: 0,
    MINoutside: 0.0,
    ANNOTATEoutput: 0,
    ATinfinity: 0,
    AVOIDold: 0,
    BESToutside: 0,
    CDDinput: 0,
    CDDoutput: 0,
    CHECKduplicates: 0,
    CHECKfrequently: 0,
    premerge_cos: 1.7976931348623157e308,
    postmerge_cos: 1.7976931348623157e308,
    DELAUNAY: 1,
    DOintersections: 0,
    DROPdim: -1,
    FLUSHprint: 0,
    FORCEoutput: 0,
    GOODpoint: 0,
    GOODpointp: 0x0000000000000000,
    GOODthreshold: 0,
    GOODvertex: 0,
    GOODvertexp: 0x0000000000000000,
    HALFspace: 0,
    ISqhullQh: 0,
    IStracing: 0,
    KEEParea: 0,
    KEEPcoplanar: 1,
    KEEPinside: 0,
    KEEPmerge: 0,
    KEEPminArea: 1.7976931348623157e308,
    MAXcoplanar: 1.7976931348623157e308,
    MAXwide: 0,
    MERGEexact: 0,
    MERGEindependent: 1,
    MERGING: 0,
    premerge_centrum: 0.0,
    postmerge_centrum: 0.0,
    MERGEpinched: 0,
    MERGEvertices: 1,
    MINvisible: 1.7976931348623157e308,
    NOnarrow: 0,
    NOnearinside: 0,
    NOpremerge: 0,
    ONLYgood: 0,
    ONLYmax: 0,
    PICKfurthest: 1,
    POSTmerge: 0,
    PREmerge: 0,
    PRINTcentrums: 0,
    PRINTcoplanar: 0,
    PRINTdim: 0,
    PRINTdots: 0,
    PRINTgood: 0,
    PRINTinner: 0,
    PRINTneighbors: 0,
    PRINTnoplanes: 0,
    PRINToptions1st: 0,
    PRINTouter: 0,
    PRINTprecision: 1,
    PRINTout: [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
    ],
    PRINTridges: 0,
    PRINTspheres: 0,
    PRINTstatistics: 0,
    PRINTsummary: 0,
    PRINTtransparent: 0,
    PROJECTdelaunay: 0,
    PROJECTinput: 0,
    RANDOMdist: 0,
    RANDOMfactor: 0.0,
    RANDOMa: 0.0,
    RANDOMb: 0.0,
    RANDOMoutside: 0,
    REPORTfreq: 0,
    REPORTfreq2: 0,
    RERUN: 0,
    ROTATErandom: -2147483648,
    SCALEinput: 0,
    SCALElast: 1,
    SETroundoff: 0,
    SKIPcheckmax: 0,
    SKIPconvex: 0,
    SPLITthresholds: 0,
    STOPadd: 0,
    STOPcone: 0,
    STOPpoint: 0,
    TESTpoints: 0,
    TESTvneighbors: 0,
    TRACElevel: 0,
    TRACElastrun: 0,
    TRACEpoint: -3,
    TRACEdist: 1.7976931348623157e308,
    TRACEmerge: 0,
    TRIangulate: 1,
    TRInormals: 0,
    UPPERdelaunay: 0,
    USEstdout: 0,
    VERIFYoutput: 0,
    VIRTUALmemory: 0,
    VORONOI: 0,
    AREAfactor: 0.0,
    DOcheckmax: 0,
    feasible_string: 0x0000000000000000,
    feasible_point: 0x0000000000000000,
    GETarea: 0,
    KEEPnearinside: 0,
    hull_dim: 0,
    input_dim: 0,
    num_points: 0,
    first_point: 0x0000000000000000,
    POINTSmalloc: 0,
    input_points: 0x0000000000000000,
    input_malloc: 0,
    qhull_command: Ok(
        "",
    ),
    qhull_commandsiz2: 0,
    rbox_command: Ok(
        "",
    ),
    qhull_options: Ok(
        "  run-id 705125642",
    ),
    qhull_optionlen: 18,
    qhull_optionsiz: 0,
    qhull_optionsiz2: 0,
    run_id: 705125642,
    VERTEXneighbors: 0,
    ZEROcentrum: 0,
    upper_threshold: 0x0000000000000000,
    lower_threshold: 0x0000000000000000,
    upper_bound: 0x0000000000000000,
    lower_bound: 0x0000000000000000,
    ANGLEround: 0.0,
    centrum_radius: 0.0,
    cos_max: 0.0,
    DISTround: 0.0,
    MAXabs_coord: 0.0,
    MAXlastcoord: 0.0,
    MAXoutside: 0.0,
    MAXsumcoord: 0.0,
    MAXwidth: -1.7976931348623157e308,
    MINdenom_1: 2.2250738585072014e-308,
    MINdenom: 0.0,
    MINdenom_1_2: 0.0,
    MINdenom_2: 0.0,
    MINlastcoord: 0.0,
    NEARzero: 0x0000000000000000,
    NEARinside: 0.0,
    ONEmerge: 0.0,
    outside_err: 1.7976931348623157e308,
    WIDEfacet: 0.0,
    NARROWhull: 0,
    qhull: [
        113,
        104,
        117,
        108,
        108,
        0,
    ],
    errexit: "...",
    jmpXtra: "...",
    restartexit: "...",
    jmpXtra2: "...",
    fin: 0x00007ea8d42038e0,
    fout: 0x00007ea8d42045c0,
    ferr: 0x000062fc557632a0,
    interior_point: 0x0000000000000000,
    normal_size: 0,
    center_size: 0,
    TEMPsize: 0,
    facet_list: 0x0000000000000000,
    facet_tail: 0x0000000000000000,
    facet_next: 0x0000000000000000,
    newfacet_list: 0x0000000000000000,
    visible_list: 0x0000000000000000,
    num_visible: 0,
    tracefacet_id: 4294967295,
    tracefacet: 0x0000000000000000,
    traceridge_id: 4294967295,
    traceridge: 0x0000000000000000,
    tracevertex_id: 4294967295,
    tracevertex: 0x0000000000000000,
    vertex_list: 0x0000000000000000,
    vertex_tail: 0x0000000000000000,
    newvertex_list: 0x0000000000000000,
    num_facets: 0,
    num_vertices: 0,
    num_outside: 0,
    num_good: 0,
    facet_id: 0,
    ridge_id: 0,
    vertex_id: 0,
    first_newfacet: 0,
    hulltime: 0,
    ALLOWrestart: 0,
    build_cnt: 0,
    CENTERtype: 0,
    furthest_id: -1,
    last_errcode: 0,
    GOODclosest: 0x0000000000000000,
    coplanar_apex: 0x0000000000000000,
    hasAreaVolume: 0,
    hasTriangulation: 0,
    isRenameVertex: 0,
    JOGGLEmax: 1.7976931348623157e308,
    maxoutdone: 0,
    max_outside: 0.0,
    max_vertex: 0.0,
    min_vertex: 0.0,
    NEWfacets: 0,
    NEWtentative: 0,
    findbestnew: 0,
    findbest_notsharp: 0,
    NOerrexit: 0,
    PRINTcradius: 0.0,
    PRINTradius: 0.0,
    POSTmerging: 0,
    printoutvar: 0,
    printoutnum: 0,
    repart_facetid: 0,
    retry_addpoint: 0,
    QHULLfinished: 0,
    totarea: 0.0,
    totvol: 0.0,
    visit_id: 0,
    vertex_visit: 0,
    WAScoplanar: 0,
    ZEROall_ok: 0,
    facet_mergeset: 0x0000000000000000,
    degen_mergeset: 0x0000000000000000,
    vertex_mergeset: 0x0000000000000000,
    hash_table: 0x0000000000000000,
    other_points: 0x0000000000000000,
    del_vertices: 0x0000000000000000,
    gm_matrix: 0x0000000000000000,
    gm_row: 0x0000000000000000,
    line: 0x0000000000000000,
    maxline: 0,
    half_space: 0x0000000000000000,
    temp_malloc: 0x0000000000000000,
    ERREXITcalled: 0,
    firstcentrum: 0,
    old_randomdist: 0,
    coplanarfacetset: 0x0000000000000000,
    last_low: 1.7976931348623157e308,
    last_high: 1.7976931348623157e308,
    last_newhigh: 1.7976931348623157e308,
    lastcpu: 0.0,
    lastfacets: 0,
    lastmerges: 0,
    lastplanes: 0,
    lastdist: 0,
    lastreport: 0,
    mergereport: 0,
    old_tempstack: 0x0000000000000000,
    ridgeoutnum: 0,
    last_random: 705125642,
    rbox_errexit: "...",
    jmpXtra3: "...",
    rbox_isinteger: 0,
    rbox_out_offset: 0.0,
    cpp_object: 0x0000000000000000,
    cpp_other: 0x0000000000000000,
    cpp_user: 0x0000000000000000,
    qhmem: qhmemT {
        BUFsize: 0,
        BUFinit: 0,
        TABLEsize: 0,
        NUMsizes: 0,
        LASTsize: 0,
        ALIGNmask: 0,
        freelists: 0x0000000000000000,
        sizetable: 0x0000000000000000,
        indextable: 0x0000000000000000,
        curbuffer: 0x0000000000000000,
        freemem: 0x0000000000000000,
        freesize: 0,
        tempstack: 0x0000000000000000,
        ferr: 0x000062fc557632a0,
        IStracing: 0,
        cntquick: 0,
        cntshort: 0,
        cntlong: 0,
        freeshort: 0,
        freelong: 0,
        totbuffer: 0,
        totdropped: 0,
        totfree: 0,
        totlong: 0,
        maxlong: 0,
        totshort: 0,
        totunused: 0,
        cntlarger: 0,
        totlarger: 0,
    },
    qhstat: "...",
}
```

</details>
this should be compared with the one produced by the reference implementation.

The reference binary (`qdelaynay_r.c`) can be run with `cat a.txt | cargo run --bin qdelaunay`.
<details>

<summary>a.txt</summary>

```
2 rbox 4
4
0 0
1 0
0 1
0.25 0.25
```

</details>

This should be just some boring work, but what really worries me is your first example that always prints `3, 3, 3` and this really needs to be investigated. (I can reproduce).

Another note: `scipy` is also using `qhull` but if I remember correctly they are just passing data as text to `qdelaunay`.

**Comment by LucaCiucci** [comment]
> Do the upper facets need filtering out - presumably by checking the direction of their normal?

`qhull` implementation is general over the dimension of the base space so it is not as simple as checking the normal direction. This should be handled by `qhull` with the "delaunay" option, but maybe the configuration is wrong...

**Comment by LucaCiucci** [comment]
Turns out that `Face::upper_delaunay` can be used to filter the relevant triangles like [this](https://github.com/LucaCiucci/qhull-rs/blob/92a476d889681e0abd3168efd5d449535f39ebc0/examples/delaunay.rs#L7). The condition is actually inverted since the paraboloid is built with a positive coefficient (for concistency with other parts of Qhull) so we have to take the bottom faces.

In the first example you wrote:
> ```rs
> for vertex in simplex.vertices().unwrap().iter() {
>     println!("{:?}", vertex.dim());
> }
> ```

But it always prints "3" since you asked for `vertex.dim()`.

@keeeal does this makes sense to you? I think this issue can be closed with #2 if it looks fine to you.

**Comment by LucaCiucci** [comment]
@keeeal I wanted to warn you that the examples using `Vertex::id` were wrong (see #3), this should be fixed in #4 that introduces `Vertex::index`
